keps/sig-network/0752-endpointslices/README.md

```diff
@@ -304,6 +301,94 @@ following well known key:
 If the Kubernetes Service has topological keys specified, the corresponding node
 labels will be copied to endpoint topology.
 
+### Converting Topology in EndpointSlice GA
+
+For beta resources that have set the topology field, the string to string map
+will be converted into an annotation on the v1 resource.
+
+The size limits of the topology field is 63B per key and per value, with a
+maximum of 126B per key-value pair. The maximum number of endpoints in an
+EndpointSlice is 1000.
+
+```
+{Max Key-Value Pairs} * ({Max Key Size} + {Max Value Size}) * {Max Endpoints}
+16 * (63 + 63) * 1000 = 2,016,000b
+```
+
+The maximum value of the combined maps in the endpoints is almost a factor of
```

Comment:
Nobody could actually achieve that, since etcd has a max 1 or 1.5 MiB for the whole resource, unless that has changed.

In practice, I expect we have 2-3 values per endpoint and most of them are the standard ones.  Rather than be lossy on some endpoints, I'd rather have something more dynamic like:

"""
We'll preserve as many key-value pairs as we can fit in 256 kB, after which we will drop kv pairs across all endpoints, iteratively, preferring to preserve standard keys.
"""

or maybe better, if we assume that all endpoints will have the same keys:

"""
We convert the list of topology maps into a `map[string][]string`.  If the total size exceeds 256 kB, we will drop keys, preferring to preserve standard keys.
"""

If we have 3 endpoints with `topology["topology.kubernetes.io/zone"]` set to "us-central1-c", "us-central1-b", "us-central1-a", then we encode it as `"topology.kubernetes.io/zone": [ "us-central1-c", "us-central1-b", "us-central1-a"].  For each key in the union of keys across all endpoints in that slice, make an index-mapped list of values.

Or, go one better.  If we assume that most clusters have a small number of
values for each topology key, something like:

```
struct {
	// Keys is a sorted list of the union of all keys in all
	// endpoints in this slice
	Keys []string

	// Values holds a list of possible values for each key in
	// the above Keys field.  The map key is an index into the
	// above Keys list and the map value is a sorted list of
	// the union of values found for that key in this slice.
	Values map[int][]string

	// Topology maps a topology key (an index into Keys) to a
	// topology value (an index into Values[tk]) for each
	// endpoint in this slice.
	Topology []map[int]int
}
```

This should ensure that every string key and string value is encoded exactly
once, and the rest are just integers.  It's still POSSIBLE to overflow this by
having different sets of keys on every endpoint, so we still need to spec that
we will drop keys and prefer standard keys, but in practice we won't see it.

```
{
	"Keys": [ "custom-thing",
	          "topology.kubernetes.io/region",
	          "topology.kubernetes.io/zone"
	],
	"Values": {
		0: [ "bar", "foo", "qux", "zrb" ],
		1: [ "us-central1" ],
		2: [ "us-central1-a", "us-central1-b", "us-central1-c"]
	},
	"Topology": [
		{ 0: 3, 1: 0, 2; 2 },
		{ 0: 2, 1: 0, 2; 0 },
		{ 0: 1, 1: 0, 2; 1 },
		{ 0: 2, 1: 0, 2; 0 },
		{ 0: 0, 1: 0, 2; 2 }
	]
}
```

You could go EVEN FURTHER and just encode it as;

```
struct {
	// Strings is a sorted list of the union of all topology 
	// keys and values in all endpoints in this slice
	Strings []string

	// Topology maps a topology key (an index into Strings) to a
	// topology value (an index into Strings) for each
	// endpoint in this slice.
	Topology []map[int]int
}
```

and get (made up):

```
{
	"Strings": [
		"bar",
		"custom-thing",
		"foo",
		"qux",
		"topology.kubernetes.io/region",
		"topology.kubernetes.io/zone",
		"us-central1",
		"us-central1-a",
		"us-central1-b",
		"us-central1-c",
		"zrb",
	],
	"Topology": [
		{ 1: 0, 3: 6, 7: 2 },
		{ 1: 2, 3: 6, 7: 0 },
		{ 1: 1, 3: 6, 7: 1 },
		{ 1: 2, 3: 6, 7: 0 },
		{ 1: 0, 3: 6, 7: 2 }
	]
```

Going much farther than that seems silly.  Just generat such a struct,
JSON-encode it, and save it as a string.  What do you think?  For the KEP, all we need to day is that we'll preserve it across round-trips, and users would have to actively try to break it, in which case we will degrade as gracefully as possible.  We could even store multiple annotations if we need more than 256KB

Reply:
:D

Reply:
@thockin This is great, thanks for the detailed explanation. A few follow up questions:

1. It looks like the [256kB limit](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L36) applies to the [total of _all_ annotations on a resource](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L47-L60). Should we try to leave some headroom here to allow for other annotations to be set? If not, how do we handle scenarios where annotations are already completely or mostly full?
2. If we're going to require consumers of the v1 API to keep this annotation up to date, some of the more optimized suggestions above feel like they could be rather difficult to keep in sync with endpoints. Do we need to provide some kind of helper that can update annotations as endpoints are removed or added?
3. Will consumers of the v1 API always be stuck with the burden of maintaining this annotation? If so, it feels like keeping the existing topology field could result in a better ux here.

Reply:
Thanks for walking through all of that @thockin! 

I agree with Rob, keeping the annotation up to date as endpoints change is likely to be difficult. We would need define a unique key per endpoint. The option of leaving the topology field does feel like a nicer UX, but one reason to remove the field was how open the field was and very little validation of what could be put into it.

Reply:
I started wondering if we really need that. The above math assumes 1000 endpoints per EndpointSlice.
But the default we are using in our controller (and recommend is 100.

If we stick to 100, then we don't seem to need to do anything - it just fits the limit.

So in order to not feet the limit people would have to:
(a) use beta API
(b) increase the EndpointSlice length above 100
(c) try to utilize the topology labels pretty aggressively

So yeah - maybe it is possible.


But going back to what Tim suggests - in all of these approaches there is a risk that something will not fit (it's super low but there is some). So we need to be prepared for dropping something in any case.
And if we need to be prepared for dropping, then can't we just drop if it's too large (dropping first the non-standard ones)?

Reply:
If a consumer wants the topology information in easy-mode, just use the v1beta1 API.  It's deprecated, so they need to get off it eventually.

> If we're going to require consumers of the v1 API to keep this annotation up to date

What are writers of v1 API doing that they need to modify this?  Especially if zone is special-cased in the v1 field.  If they consume it in v1beta1, zone gets written into the map.

If you are a proper v1 consumer, you either totally ignore this map (or log "v1beta1 topoloogy is deprecated - ignoring it") or you write a function to decode it.  It's not a difficult encoding.  v1 clients should never write to it.

Packing an annotation with a hundred duplicates of the same string just feels silly :)

Reply:
oh, I guess if a v1 client wanted to add endpoints they would need to preserve this encoding and insert a new one.  I still don't see that as too bad.

Reply:
I think removing endpoints is the worst part since instead of appending to an annotation you'd have to move everything around within the annotation. Especially if this is something we're stuck with _forever_, I think it's less bad just to keep the topology field. On the other hand, if this could be a short term requirement for a couple releases or something I think it's fine.

Reply:
> v1 clients in a mixed mode need something like an unpack function and a mapping per endpoint -> annotation payload. If they edit a slice, they need to re-assemble the annotation. That doesn't sound so bad?
I think Tim's [comment](https://github.com/kubernetes/enhancements/pull/2367#discussion_r568190121) in a different comment thread will satisfy our concerns. I think we just need to provide clear documentation informing people who are using a v1 client that they will need to unpack and update the topology annotation when they make edits to the endpoints. It does put some onus on the user but I think that's is okay. Users already will need to unpack the annotation to be able to use.

Reply:
I have updated the KEP so it says a conversion functionality will be provided and due to space requirements, we will drop KV pairs starting with the non-standard ones. I provided a summary of the option Tim outlined above as a possible space optimization. I can make this more detailed if that would be more helpful

Reply:
Thanks for the updates here! As I've thought about it more, I've realized it really isn't that bad. EndpointSlices are generally managed by a single component so I think it's unlikely that annotations will be an issue here. Agree that this is our best way to remove this field.